src/xtgeo/grid3d/_grid_import_xtgcpgeom.py

```diff
@@ -48,72 +86,34 @@ def import_xtgcpgeom(
     ncoord = nncol * nnrow * 6
     nzcorn = nncol * nnrow * nnlay * 4
     nactnum = ncol * nrow * nlay
-
     # read numpy arrays from file
-    coordsv = xsys.npfromfile(
+    self._coordsv = xsys.npfromfile(
         mfile.file, dtype=dtype_coordsv, count=ncoord, offset=offset, mmap=mmap
     )
     newoffset = offset + ncoord * coordfmt
-    zcornsv = xsys.npfromfile(
+    self._zcornsv = xsys.npfromfile(
         mfile.file, dtype=dtype_zcornsv, count=nzcorn, offset=newoffset, mmap=mmap
     )
     newoffset += nzcorn * zcornfmt
-    actnumsv = xsys.npfromfile(
+    self._actnumsv = xsys.npfromfile(
         mfile.file, dtype=dtype_actnumv, count=nactnum, offset=newoffset, mmap=mmap
     )
     newoffset += nactnum * actnumfmt
 
     # read metadata which will be at position offet + nfloat*narr +13
     pos = newoffset + 13
-
     with open(mfile.file, "rb") as fhandle:
         fhandle.seek(pos)
         jmeta = fhandle.read().decode()
 
     meta = json.loads(jmeta, object_pairs_hook=OrderedDict)
-    req = meta["_required_"]
-
-    # meta _optional_ *may* contain xshift, xscale etc which in case must be taken
-    # into account
-    opt = meta.get("_optional_", None)
-    if opt:
-        if {"xshift", "xscale"}.issubset(opt):
-            shi = opt["xshift"]
-            sca = opt["xscale"]
-            coordsv[0::3] = np.where(shi != 0, coordsv[0::3] + shi, coordsv[0::3])
-            coordsv[0::3] = np.where(sca != 1, coordsv[0::3] * sca, coordsv[0::3])
-        if {"yshift", "yscale"}.issubset(opt):
-            shi = opt["yshift"]
-            sca = opt["yscale"]
-            coordsv[1::3] = np.where(shi != 0, coordsv[1::3] + shi, coordsv[1::3])
-            coordsv[1::3] = np.where(sca != 1, coordsv[1::3] * sca, coordsv[1::3])
-        if {"zshift", "zscale"}.issubset(opt):
-            shi = opt["zshift"]
-            sca = opt["zscale"]
-            coordsv[2::3] = np.where(shi != 0, coordsv[2::3] + shi, coordsv[2::3])
-            coordsv[2::3] = np.where(sca != 1, coordsv[2::3] * sca, coordsv[2::3])
-            zcornsv = (zcornsv + shi) * sca
-
-    self._coordsv = coordsv.reshape((nncol, nnrow, 6)).astype(np.float64)
-    self._zcornsv = zcornsv.reshape((nncol, nnrow, nnlay, 4)).astype(np.float32)
-    self._actnumsv = actnumsv.reshape((ncol, nrow, nlay)).astype(np.int32)
 
-    reqattrs = xtgeo.MetaDataCPGeometry.REQUIRED
-
-    for myattr in reqattrs:
-        if "subgrid" in myattr:
-            self.set_subgrids(reqattrs["subgrids"])
-        else:
-            setattr(self, "_" + myattr, req[myattr])
-
-    self._metadata.required = self
+    handle_metadata(self, meta, ncol, nrow, nlay)
```

Comment:
Would it be possible to refactor the `import_xtgcpgeom` to simply return the three variables instead of passing on self? Here we're just taking this awkward setting of private parameters just further down.

Would really appreciate it at least :)

Reply:
Going away from self is adressed in https://github.com/equinor/xtgeo/pull/655 :)